---a/net/minecraft/block/BlockPortal.java
+++b/net/minecraft/block/BlockPortal.java
@@ -1,6 +1,10 @@
 package net.minecraft.block;
 
+import java.util.Collection;
+import java.util.HashSet;
 import java.util.Random;
+
 import net.minecraft.block.material.Material;
 import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyEnum;
@@ -8,405 +8,308 @@
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.Entity;
 import net.minecraft.init.Blocks;
-import net.minecraft.item.Item;
 import net.minecraft.item.ItemMonsterPlacer;
 import net.minecraft.util.AxisAlignedBB;
 import net.minecraft.util.BlockPos;
 import net.minecraft.util.EnumFacing;
-import net.minecraft.util.EnumParticleTypes;
-import net.minecraft.util.EnumWorldBlockLayer;
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class BlockPortal extends BlockBreakable
-{
-    public static final PropertyEnum AXIS = PropertyEnum.create("axis", EnumFacing.Axis.class, new EnumFacing.Axis[] {EnumFacing.Axis.X, EnumFacing.Axis.Z});
+import org.bukkit.Location;
+import org.bukkit.event.entity.EntityPortalEnterEvent;
+import org.bukkit.event.world.PortalCreateEvent;
+
+public class BlockPortal extends BlockBreakable {
+
+    public static final PropertyEnum AXIS = PropertyEnum.create("axis", EnumFacing.Axis.class, new EnumFacing.Axis[]{EnumFacing.Axis.X, EnumFacing.Axis.Z});
     private static final String __OBFID = "CL_00000284";
 
-    public BlockPortal()
-    {
+    public BlockPortal() {
         super(Material.portal, false);
         this.setDefaultState(this.blockState.getBaseState().withProperty(AXIS, EnumFacing.Axis.X));
         this.setTickRandomly(true);
     }
 
-    public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand)
-    {
+    public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand) {
         super.updateTick(worldIn, pos, state, rand);
 
-        if (worldIn.provider.isSurfaceWorld() && worldIn.getGameRules().getGameRuleBooleanValue("doMobSpawning") && rand.nextInt(2000) < worldIn.getDifficulty().getDifficultyId())
-        {
+        if (worldIn.spigotConfig.enableZombiePigmenPortalSpawns && worldIn.provider.isSurfaceWorld() && worldIn.getGameRules().getGameRuleBooleanValue("doMobSpawning") && rand.nextInt(2000) < worldIn.getDifficulty().getDifficultyId()) { // Spigot        {
             int i = pos.getY();
             BlockPos blockpos1;
 
-            for (blockpos1 = pos; !World.doesBlockHaveSolidTopSurface(worldIn, blockpos1) && blockpos1.getY() > 0; blockpos1 = blockpos1.down())
-            {
+            for (blockpos1 = pos; !World.doesBlockHaveSolidTopSurface(worldIn, blockpos1) && blockpos1.getY() > 0; blockpos1 = blockpos1.down()) {
                 ;
             }
 
-            if (i > 0 && !worldIn.getBlockState(blockpos1.up()).getBlock().isNormalCube())
-            {
-                Entity entity = ItemMonsterPlacer.spawnCreature(worldIn, 57, (double)blockpos1.getX() + 0.5D, (double)blockpos1.getY() + 1.1D, (double)blockpos1.getZ() + 0.5D);
+            if (i > 0 && !worldIn.getBlockState(blockpos1.up()).getBlock().isNormalCube()) {
+                // CraftBukkit - set spawn reason to NETHER_PORTAL
+                Entity entity = ItemMonsterPlacer.spawnCreature(worldIn, 57, (double) blockpos1.getX() + 0.5D, (double) blockpos1.getY() + 1.1D, (double) blockpos1.getZ() + 0.5D, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NETHER_PORTAL);
 
-                if (entity != null)
-                {
+                if (entity != null) {
                     entity.timeUntilPortal = entity.getPortalCooldown();
                 }
             }
         }
     }
 
-    public AxisAlignedBB getCollisionBoundingBox(World worldIn, BlockPos pos, IBlockState state)
-    {
+    public AxisAlignedBB getCollisionBoundingBox(World worldIn, BlockPos pos, IBlockState state) {
         return null;
     }
 
-    public void setBlockBoundsBasedOnState(IBlockAccess worldIn, BlockPos pos)
-    {
-        EnumFacing.Axis axis = (EnumFacing.Axis)worldIn.getBlockState(pos).getValue(AXIS);
+    public void setBlockBoundsBasedOnState(IBlockAccess worldIn, BlockPos pos) {
+        EnumFacing.Axis axis = (EnumFacing.Axis) worldIn.getBlockState(pos).getValue(AXIS);
         float f = 0.125F;
         float f1 = 0.125F;
 
-        if (axis == EnumFacing.Axis.X)
-        {
+        if (axis == EnumFacing.Axis.X) {
             f = 0.5F;
         }
 
-        if (axis == EnumFacing.Axis.Z)
-        {
+        if (axis == EnumFacing.Axis.Z) {
             f1 = 0.5F;
         }
 
         this.setBlockBounds(0.5F - f, 0.0F, 0.5F - f1, 0.5F + f, 1.0F, 0.5F + f1);
     }
 
-    public static int getMetaForAxis(EnumFacing.Axis axis)
-    {
+    public static int getMetaForAxis(EnumFacing.Axis axis) {
         return axis == EnumFacing.Axis.X ? 1 : (axis == EnumFacing.Axis.Z ? 2 : 0);
     }
 
-    public boolean isFullCube()
-    {
+    public boolean isFullCube() {
         return false;
     }
 
-    public boolean func_176548_d(World worldIn, BlockPos p_176548_2_)
-    {
+    public boolean func_176548_d(World worldIn, BlockPos p_176548_2_) {
         BlockPortal.Size size = new BlockPortal.Size(worldIn, p_176548_2_, EnumFacing.Axis.X);
 
-        if (size.func_150860_b() && size.field_150864_e == 0)
-        {
-            size.func_150859_c();
-            return true;
-        }
-        else
-        {
+        if (size.func_150860_b() && size.field_150864_e == 0) {
+            // CraftBukkit start -- Return portal creator
+            return size.func_150859_c();
+            // return true;
+        } else {
             BlockPortal.Size size1 = new BlockPortal.Size(worldIn, p_176548_2_, EnumFacing.Axis.Z);
 
-            if (size1.func_150860_b() && size1.field_150864_e == 0)
-            {
-                size1.func_150859_c();
-                return true;
-            }
-            else
-            {
+            if (size1.func_150860_b() && size1.field_150864_e == 0) {
+                return size1.func_150859_c();
+                // return true;
+                // CraftBukkit end
+            } else {
                 return false;
             }
         }
     }
 
-    public void onNeighborBlockChange(World worldIn, BlockPos pos, IBlockState state, Block neighborBlock)
-    {
-        EnumFacing.Axis axis = (EnumFacing.Axis)state.getValue(AXIS);
+    public void onNeighborBlockChange(World worldIn, BlockPos pos, IBlockState state, Block neighborBlock) {
+        EnumFacing.Axis axis = (EnumFacing.Axis) state.getValue(AXIS);
         BlockPortal.Size size;
 
-        if (axis == EnumFacing.Axis.X)
-        {
+        if (axis == EnumFacing.Axis.X) {
             size = new BlockPortal.Size(worldIn, pos, EnumFacing.Axis.X);
 
-            if (!size.func_150860_b() || size.field_150864_e < size.field_150868_h * size.field_150862_g)
-            {
+            if (!size.func_150860_b() || size.field_150864_e < size.field_150868_h * size.field_150862_g) {
                 worldIn.setBlockState(pos, Blocks.air.getDefaultState());
             }
-        }
-        else if (axis == EnumFacing.Axis.Z)
-        {
+        } else if (axis == EnumFacing.Axis.Z) {
             size = new BlockPortal.Size(worldIn, pos, EnumFacing.Axis.Z);
 
-            if (!size.func_150860_b() || size.field_150864_e < size.field_150868_h * size.field_150862_g)
-            {
+            if (!size.func_150860_b() || size.field_150864_e < size.field_150868_h * size.field_150862_g) {
                 worldIn.setBlockState(pos, Blocks.air.getDefaultState());
             }
         }
     }
 
-    @SideOnly(Side.CLIENT)
-    public boolean shouldSideBeRendered(IBlockAccess worldIn, BlockPos pos, EnumFacing side)
-    {
-        EnumFacing.Axis axis = null;
-        IBlockState iblockstate = worldIn.getBlockState(pos);
-
-        if (worldIn.getBlockState(pos).getBlock() == this)
-        {
-            axis = (EnumFacing.Axis)iblockstate.getValue(AXIS);
-
-            if (axis == null)
-            {
-                return false;
-            }
-
-            if (axis == EnumFacing.Axis.Z && side != EnumFacing.EAST && side != EnumFacing.WEST)
-            {
-                return false;
-            }
-
-            if (axis == EnumFacing.Axis.X && side != EnumFacing.SOUTH && side != EnumFacing.NORTH)
-            {
-                return false;
-            }
-        }
-
-        boolean flag = worldIn.getBlockState(pos.west()).getBlock() == this && worldIn.getBlockState(pos.west(2)).getBlock() != this;
-        boolean flag1 = worldIn.getBlockState(pos.east()).getBlock() == this && worldIn.getBlockState(pos.east(2)).getBlock() != this;
-        boolean flag2 = worldIn.getBlockState(pos.north()).getBlock() == this && worldIn.getBlockState(pos.north(2)).getBlock() != this;
-        boolean flag3 = worldIn.getBlockState(pos.south()).getBlock() == this && worldIn.getBlockState(pos.south(2)).getBlock() != this;
-        boolean flag4 = flag || flag1 || axis == EnumFacing.Axis.X;
-        boolean flag5 = flag2 || flag3 || axis == EnumFacing.Axis.Z;
-        return flag4 && side == EnumFacing.WEST ? true : (flag4 && side == EnumFacing.EAST ? true : (flag5 && side == EnumFacing.NORTH ? true : flag5 && side == EnumFacing.SOUTH));
-    }
-
-    public int quantityDropped(Random random)
-    {
+    public int quantityDropped(Random random) {
         return 0;
     }
 
-    public void onEntityCollidedWithBlock(World worldIn, BlockPos pos, IBlockState state, Entity entityIn)
-    {
-        if (entityIn.ridingEntity == null && entityIn.riddenByEntity == null)
-        {
+    public void onEntityCollidedWithBlock(World worldIn, BlockPos pos, IBlockState state, Entity entityIn) {
+        if (entityIn.ridingEntity == null && entityIn.riddenByEntity == null) {
+            // CraftBukkit start - Entity in portal
+            EntityPortalEnterEvent event = new EntityPortalEnterEvent(entityIn.getBukkitEntity(), new Location(worldIn.getWorld(), pos.getX(), pos.getY(), pos.getZ()));
+            worldIn.getServer().getPluginManager().callEvent(event);
+            // CraftBukkit end
             entityIn.setInPortal();
         }
     }
 
-    public IBlockState getStateFromMeta(int meta)
-    {
+    public IBlockState getStateFromMeta(int meta) {
         return this.getDefaultState().withProperty(AXIS, (meta & 3) == 2 ? EnumFacing.Axis.Z : EnumFacing.Axis.X);
     }
 
-    @SideOnly(Side.CLIENT)
-    public EnumWorldBlockLayer getBlockLayer()
-    {
-        return EnumWorldBlockLayer.TRANSLUCENT;
+    public int getMetaFromState(IBlockState state) {
+        return getMetaForAxis((EnumFacing.Axis) state.getValue(AXIS));
     }
 
-    @SideOnly(Side.CLIENT)
-    public void randomDisplayTick(World worldIn, BlockPos pos, IBlockState state, Random rand)
-    {
-        if (rand.nextInt(100) == 0)
-        {
-            worldIn.playSound((double)pos.getX() + 0.5D, (double)pos.getY() + 0.5D, (double)pos.getZ() + 0.5D, "portal.portal", 0.5F, rand.nextFloat() * 0.4F + 0.8F, false);
-        }
+    protected BlockState createBlockState() {
+        return new BlockState(this, new IProperty[]{AXIS});
+    }
 
-        for (int i = 0; i < 4; ++i)
-        {
-            double d0 = (double)((float)pos.getX() + rand.nextFloat());
-            double d1 = (double)((float)pos.getY() + rand.nextFloat());
-            double d2 = (double)((float)pos.getZ() + rand.nextFloat());
-            double d3 = ((double)rand.nextFloat() - 0.5D) * 0.5D;
-            double d4 = ((double)rand.nextFloat() - 0.5D) * 0.5D;
-            double d5 = ((double)rand.nextFloat() - 0.5D) * 0.5D;
-            int j = rand.nextInt(2) * 2 - 1;
+    public static class Size {
 
-            if (worldIn.getBlockState(pos.west()).getBlock() != this && worldIn.getBlockState(pos.east()).getBlock() != this)
-            {
-                d0 = (double)pos.getX() + 0.5D + 0.25D * (double)j;
-                d3 = (double)(rand.nextFloat() * 2.0F * (float)j);
-            }
-            else
-            {
-                d2 = (double)pos.getZ() + 0.5D + 0.25D * (double)j;
-                d5 = (double)(rand.nextFloat() * 2.0F * (float)j);
-            }
+        private final World world;
+        private final EnumFacing.Axis axis;
+        private final EnumFacing field_150866_c;
+        private final EnumFacing field_150863_d;
+        private int field_150864_e = 0;
+        private BlockPos field_150861_f;
+        private int field_150862_g;
+        private int field_150868_h;
+        Collection blocks = new HashSet(); // CraftBukkit - add field
+        private static final String __OBFID = "CL_00000285";
 
-            worldIn.spawnParticle(EnumParticleTypes.PORTAL, d0, d1, d2, d3, d4, d5, new int[0]);
-        }
-    }
+        public Size(World worldIn, BlockPos p_i45694_2_, EnumFacing.Axis p_i45694_3_) {
+            this.world = worldIn;
+            this.axis = p_i45694_3_;
 
-    public int getMetaFromState(IBlockState state)
-    {
-        return getMetaForAxis((EnumFacing.Axis)state.getValue(AXIS));
-    }
+            if (p_i45694_3_ == EnumFacing.Axis.X) {
+                this.field_150863_d = EnumFacing.EAST;
+                this.field_150866_c = EnumFacing.WEST;
+            } else {
+                this.field_150863_d = EnumFacing.NORTH;
+                this.field_150866_c = EnumFacing.SOUTH;
+            }
 
-    @SideOnly(Side.CLIENT)
-    public Item getItem(World worldIn, BlockPos pos)
-    {
-        return null;
-    }
+            for (BlockPos blockpos1 = p_i45694_2_; p_i45694_2_.getY() > blockpos1.getY() - 21 && p_i45694_2_.getY() > 0 && this.func_150857_a(worldIn.getBlockState(p_i45694_2_.down()).getBlock()); p_i45694_2_ = p_i45694_2_.down()) {
+                ;
+            }
 
-    protected BlockState createBlockState()
-    {
-        return new BlockState(this, new IProperty[] {AXIS});
-    }
+            int i = this.func_180120_a(p_i45694_2_, this.field_150863_d) - 1;
 
-    public static class Size
-        {
-            private final World world;
-            private final EnumFacing.Axis axis;
-            private final EnumFacing field_150866_c;
-            private final EnumFacing field_150863_d;
-            private int field_150864_e = 0;
-            private BlockPos field_150861_f;
-            private int field_150862_g;
-            private int field_150868_h;
-            private static final String __OBFID = "CL_00000285";
+            if (i >= 0) {
+                this.field_150861_f = p_i45694_2_.offset(this.field_150863_d, i);
+                this.field_150868_h = this.func_180120_a(this.field_150861_f, this.field_150866_c);
 
-            public Size(World worldIn, BlockPos p_i45694_2_, EnumFacing.Axis p_i45694_3_)
-            {
-                this.world = worldIn;
-                this.axis = p_i45694_3_;
-
-                if (p_i45694_3_ == EnumFacing.Axis.X)
-                {
-                    this.field_150863_d = EnumFacing.EAST;
-                    this.field_150866_c = EnumFacing.WEST;
+                if (this.field_150868_h < 2 || this.field_150868_h > 21) {
+                    this.field_150861_f = null;
+                    this.field_150868_h = 0;
                 }
-                else
-                {
-                    this.field_150863_d = EnumFacing.NORTH;
-                    this.field_150866_c = EnumFacing.SOUTH;
-                }
+            }
 
-                for (BlockPos blockpos1 = p_i45694_2_; p_i45694_2_.getY() > blockpos1.getY() - 21 && p_i45694_2_.getY() > 0 && this.func_150857_a(worldIn.getBlockState(p_i45694_2_.down()).getBlock()); p_i45694_2_ = p_i45694_2_.down())
-                {
-                    ;
-                }
+            if (this.field_150861_f != null) {
+                this.field_150862_g = this.func_150858_a();
+            }
+        }
 
-                int i = this.func_180120_a(p_i45694_2_, this.field_150863_d) - 1;
+        protected int func_180120_a(BlockPos p_180120_1_, EnumFacing p_180120_2_) {
+            int i;
 
-                if (i >= 0)
-                {
-                    this.field_150861_f = p_i45694_2_.offset(this.field_150863_d, i);
-                    this.field_150868_h = this.func_180120_a(this.field_150861_f, this.field_150866_c);
+            for (i = 0; i < 22; ++i) {
+                BlockPos blockpos1 = p_180120_1_.offset(p_180120_2_, i);
 
-                    if (this.field_150868_h < 2 || this.field_150868_h > 21)
-                    {
-                        this.field_150861_f = null;
-                        this.field_150868_h = 0;
-                    }
+                if (!this.func_150857_a(this.world.getBlockState(blockpos1).getBlock()) || this.world.getBlockState(blockpos1.down()).getBlock() != Blocks.obsidian) {
+                    break;
                 }
-
-                if (this.field_150861_f != null)
-                {
-                    this.field_150862_g = this.func_150858_a();
-                }
             }
 
-            protected int func_180120_a(BlockPos p_180120_1_, EnumFacing p_180120_2_)
-            {
-                int i;
+            Block block = this.world.getBlockState(p_180120_1_.offset(p_180120_2_, i)).getBlock();
+            return block == Blocks.obsidian ? i : 0;
+        }
 
-                for (i = 0; i < 22; ++i)
-                {
-                    BlockPos blockpos1 = p_180120_1_.offset(p_180120_2_, i);
+        protected int func_150858_a() {
+            // CraftBukkit start
+            this.blocks.clear();
+            org.bukkit.World bworld = this.world.getWorld();
+            // CraftBukkit end
+            int i;
+            label56:
 
-                    if (!this.func_150857_a(this.world.getBlockState(blockpos1).getBlock()) || this.world.getBlockState(blockpos1.down()).getBlock() != Blocks.obsidian)
-                    {
-                        break;
+            for (this.field_150862_g = 0; this.field_150862_g < 21; ++this.field_150862_g) {
+                for (i = 0; i < this.field_150868_h; ++i) {
+                    BlockPos blockpos = this.field_150861_f.offset(this.field_150866_c, i).up(this.field_150862_g);
+                    Block block = this.world.getBlockState(blockpos).getBlock();
+
+                    if (!this.func_150857_a(block)) {
+                        break label56;
                     }
-                }
 
-                Block block = this.world.getBlockState(p_180120_1_.offset(p_180120_2_, i)).getBlock();
-                return block == Blocks.obsidian ? i : 0;
-            }
+                    if (block == Blocks.portal) {
+                        ++this.field_150864_e;
+                    }
 
-            protected int func_150858_a()
-            {
-                int i;
-                label56:
+                    if (i == 0) {
+                        block = this.world.getBlockState(blockpos.offset(this.field_150863_d)).getBlock();
 
-                for (this.field_150862_g = 0; this.field_150862_g < 21; ++this.field_150862_g)
-                {
-                    for (i = 0; i < this.field_150868_h; ++i)
-                    {
-                        BlockPos blockpos = this.field_150861_f.offset(this.field_150866_c, i).up(this.field_150862_g);
-                        Block block = this.world.getBlockState(blockpos).getBlock();
-
-                        if (!this.func_150857_a(block))
-                        {
+                        if (block != Blocks.obsidian) {
                             break label56;
+                            // CraftBukkit start - add the block to our list
+                        } else {
+                            BlockPos pos = blockpos.offset(this.field_150863_d);
+                            blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
+                            // CraftBukkit end
                         }
+                    } else if (i == this.field_150868_h - 1) {
+                        block = this.world.getBlockState(blockpos.offset(this.field_150866_c)).getBlock();
 
-                        if (block == Blocks.portal)
-                        {
-                            ++this.field_150864_e;
+                        if (block != Blocks.obsidian) {
+                            break label56;
+                            // CraftBukkit start - add the block to our list
+                        } else {
+                            BlockPos pos = blockpos.offset(this.field_150863_d);
+                            blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
+                            // CraftBukkit end
                         }
-
-                        if (i == 0)
-                        {
-                            block = this.world.getBlockState(blockpos.offset(this.field_150863_d)).getBlock();
-
-                            if (block != Blocks.obsidian)
-                            {
-                                break label56;
-                            }
-                        }
-                        else if (i == this.field_150868_h - 1)
-                        {
-                            block = this.world.getBlockState(blockpos.offset(this.field_150866_c)).getBlock();
-
-                            if (block != Blocks.obsidian)
-                            {
-                                break label56;
-                            }
-                        }
                     }
                 }
+            }
 
-                for (i = 0; i < this.field_150868_h; ++i)
-                {
-                    if (this.world.getBlockState(this.field_150861_f.offset(this.field_150866_c, i).up(this.field_150862_g)).getBlock() != Blocks.obsidian)
-                    {
-                        this.field_150862_g = 0;
-                        break;
-                    }
-                }
-
-                if (this.field_150862_g <= 21 && this.field_150862_g >= 3)
-                {
-                    return this.field_150862_g;
-                }
-                else
-                {
-                    this.field_150861_f = null;
-                    this.field_150868_h = 0;
+            for (i = 0; i < this.field_150868_h; ++i) {
+                if (this.world.getBlockState(this.field_150861_f.offset(this.field_150866_c, i).up(this.field_150862_g)).getBlock() != Blocks.obsidian) {
                     this.field_150862_g = 0;
-                    return 0;
+                    break;
+                    // CraftBukkit start - add the block to our list
+                } else {
+                    BlockPos pos = this.field_150861_f.offset(this.field_150866_c, i).up(this.field_150862_g);
+                    blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
+                    // CraftBukkit end
                 }
             }
 
-            protected boolean func_150857_a(Block p_150857_1_)
-            {
-                return p_150857_1_.blockMaterial == Material.air || p_150857_1_ == Blocks.fire || p_150857_1_ == Blocks.portal;
+            if (this.field_150862_g <= 21 && this.field_150862_g >= 3) {
+                return this.field_150862_g;
+            } else {
+                this.field_150861_f = null;
+                this.field_150868_h = 0;
+                this.field_150862_g = 0;
+                return 0;
             }
+        }
 
-            public boolean func_150860_b()
-            {
-                return this.field_150861_f != null && this.field_150868_h >= 2 && this.field_150868_h <= 21 && this.field_150862_g >= 3 && this.field_150862_g <= 21;
+        protected boolean func_150857_a(Block p_150857_1_) {
+            return p_150857_1_.blockMaterial == Material.air || p_150857_1_ == Blocks.fire || p_150857_1_ == Blocks.portal;
+        }
+
+        public boolean func_150860_b() {
+            return this.field_150861_f != null && this.field_150868_h >= 2 && this.field_150868_h <= 21 && this.field_150862_g >= 3 && this.field_150862_g <= 21;
+        }
+
+        // CraftBukkit start - return boolean
+        public boolean func_150859_c() {
+            org.bukkit.World bworld = this.world.getWorld();
+
+            // Copy below for loop
+            for (int i = 0; i < this.field_150868_h; ++i) {
+                BlockPos blockposition = this.field_150861_f.offset(this.field_150866_c, i);
+
+                for (int j = 0; j < this.field_150862_g; ++j) {
+                    BlockPos pos = blockposition.up(j);
+                    blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
+                }
             }
 
-            public void func_150859_c()
-            {
-                for (int i = 0; i < this.field_150868_h; ++i)
-                {
-                    BlockPos blockpos = this.field_150861_f.offset(this.field_150866_c, i);
+            PortalCreateEvent event = new PortalCreateEvent(blocks, bworld, PortalCreateEvent.CreateReason.FIRE);
+            this.world.getServer().getPluginManager().callEvent(event);
 
-                    for (int j = 0; j < this.field_150862_g; ++j)
-                    {
-                        this.world.setBlockState(blockpos.up(j), Blocks.portal.getDefaultState().withProperty(BlockPortal.AXIS, this.axis), 2);
-                    }
+            if (event.isCancelled()) {
+                return false;
+            }
+            // CraftBukkit end
+            for (int i = 0; i < this.field_150868_h; ++i) {
+                BlockPos blockpos = this.field_150861_f.offset(this.field_150866_c, i);
+
+                for (int j = 0; j < this.field_150862_g; ++j) {
+                    this.world.setBlockState(blockpos.up(j), Blocks.portal.getDefaultState().withProperty(BlockPortal.AXIS, this.axis), 2);
                 }
             }
+            return true; // Craft Bukkit
         }
-}
+    }
+}
